[ iOS simulator ] 2X TestWebKitAPI.DragAndDropTests.ExternalSourceDataTransferItemGet (API-Tests) are constant timeouts

aprotyas opened the pull request:
#### bef4c7f2d0b60199d448ed763bc2edc554c55594
<pre>
[ iOS simulator ] 2X TestWebKitAPI.DragAndDropTests.ExternalSourceDataTransferItemGet (API-Tests) are constant timeouts
<a href="https://bugs.webkit.org/show_bug.cgi?id=256746">https://bugs.webkit.org/show_bug.cgi?id=256746</a>
rdar://109222780

Reviewed by Wenson Hsieh.

The API tests TestWebKitAPI.DragAndDropTests.ExternalSourceDataTransferItemGet[Folder,PlainTextFile]AsEntry
consistently time out when running on an iOS simulator simulating iPhone
12. These timeouts have surfaced recently given we bumped the default
iPhone simulator to iPhone 12 in 256495@main. When running on iPhone 12,
the change in screen resolution meant that some of our hardcoded drag
gestures became out-of-place, causing unwarranted behavior and leading
to the timeouts.

This commit hardens the test cases against similar changes in the future
by introducing a way to start a drag over an element in the DOM and
ending it at another element in the DOM. Doing so required imbibing
`TestWKWebView` with a `getElementMidpoint` method, which takes the
selector for a DOM element and returns its midpoint, and
`DragAndDropSimulator` with a `runFromElement:toElement:` method, which
is similar in spirit to the `runFrom:to:` except the drag source and
destination are now specified as DOM element selectors, rather than
locations.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm: Added.
(-[DragAndDropSimulator runFromElement:toElement:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView getElementMidpoint:]):

Signed-off-by: Abrar Rahman Protyasha &lt;a_protyasha@apple.com&gt;
Canonical link: <a href="https://commits.webkit.org/264080@main">https://commits.webkit.org/264080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9f8d6940b38b36eafdacb26de28cd7d1d62a331

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6641 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7039 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8232 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7928 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6805 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9815 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6759 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6080 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8319 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6009 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/6105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8773 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6575 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5386 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5971 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1568 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6344 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
```diff
@@ -420,6 +420,23 @@ - (void)simulateAllTouchesCanceled:(NSNotification *)notification
         [[_webView dragInteractionDelegate] dragInteraction:[_webView dragInteraction] session:_dragSession.get() didEndWithOperation:UIDropOperationCancel];
 }
 
+- (void)runFromElement:(NSString *)startID toElement:(NSString *)endID
```
Nit - maybe rename these to `startSelector` and `endSelector`?

aprotyas replied:
Renamed in https://github.com/WebKit/WebKit/commit/e92cb7bf8985c7c8b09a62d159bf24619ec278dd

whsieh commented on Tools/TestWebKitAPI/mac/DragAndDropSimulatorMac.mm:
```diff
@@ -151,6 +151,23 @@ - (double)initialProgressForMouseDrag
     return !totalDistance ? 1 : std::min<double>(1, initialMouseDragDistance / totalDistance);
 }
 
+- (void)runFromElement:(NSString *)startID toElement:(NSString *)endID
```
We should probably also put this implementation in a single place, instead of copying it in two places — e.g. `DragAndDropSimulator.mm`. You can probably use an anonymous category:

```
@implementation DragAndDropSimulator ()
…
@end
```
...for that.

aprotyas replied:
I've put the implementation in a single place (`cocoa/DragAndDropSimulator.mm`), but couldn't get the anonymous category approach to work. What do you think of the current solution in https://github.com/WebKit/WebKit/pull/13844/commits/e92cb7bf8985c7c8b09a62d159bf24619ec278dd? And/or do you know why this didn't work?

```obj-c
// In DragAndDropSimulator.mm

@interface DragAndDropSimulator ()

- (void)runFromElement:(NSString *)startSelector toElement:(NSString *)endSelector;

@end


@implementation DragAndDropSimulator

- (void)runFromElement:(NSString *)startSelector toElement:(NSString *)endSelector
{
    // ...
}

@end
```

The compiler can't find `runFromElement:toElement:` in `DragAndDropTestsIOS.mm` and also suggests that `@implementation DragAndDropSimulator` in `DragAndDropSimulator.mm` should implement the full interface `@interface DragAndDropSimulator : NSObject<WKUIDelegatePrivate, _WKInputDelegate>` described in `DragAndDropSimulator.h`.

<details>
<summary>Build error from the anonymous category approach</summary>

```
TestWebKitAPILibrary CompileC DragAndDropTestsIOS.o
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:1214:20: error: instance method '-runFromElement:toElement:' not found (return type defaults to 'id') [-Werror,-Wobjc-method-access]
        [simulator runFromElement:@"#output" toElement:@"#droparea"];
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:31:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:79:12: note: receiver is instance of class declared here
@interface DragAndDropSimulator : NSObject<WKUIDelegatePrivate, _WKInputDelegate>
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:1244:20: error: instance method '-runFromElement:toElement:' not found (return type defaults to 'id') [-Werror,-Wobjc-method-access]
        [simulator runFromElement:@"#output" toElement:@"#droparea"];
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:31:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:79:12: note: receiver is instance of class declared here
@interface DragAndDropSimulator : NSObject<WKUIDelegatePrivate, _WKInputDelegate>
TestWebKitAPILibrary CompileC UnifiedSource9-mm.o
TestWebKitAPILibrary CompileC UnifiedSource39-mm.o
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:81:1: note: method 'initWithWebViewFrame:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:configuration:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:82:1: note: method 'initWithWebViewFrame:configuration:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:84:1: note: method 'runFrom:to:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'endDataTransfer' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:85:1: note: method 'endDataTransfer' declared here
- (void)endDataTransfer;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'clearExternalDragInformation' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:86:1: note: method 'clearExternalDragInformation' declared here
- (void)clearExternalDragInformation;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebView:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:94:1: note: method 'initWithWebView:' declared here
- (instancetype)initWithWebView:(TestWKWebView *)webView;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:additionalItemRequestLocations:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:95:1: note: method 'runFrom:to:additionalItemRequestLocations:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation additionalItemRequestLocations:(ProgressToCGPointValueMap)additionalItemRequestLocations;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'ensureInputSession' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:96:1: note: method 'ensureInputSession' declared here
- (void)ensureInputSession;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'setExternalItemProviders:defaultDropPreviews:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:97:1: note: method 'setExternalItemProviders:defaultDropPreviews:' declared here
- (void)setExternalItemProviders:(NSArray<NSItemProvider *> *)itemProviders defaultDropPreviews:(NSArray<UITargetedDragPreview *> *)previews;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'containsDraggedType:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:145:1: note: method 'containsDraggedType:' declared here
- (BOOL)containsDraggedType:(NSString *)type;
^
TestWebKitAPILibrary CompileC DragAndDropSimulator.o
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:81:1: note: method 'initWithWebViewFrame:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:configuration:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:82:1: note: method 'initWithWebViewFrame:configuration:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:84:1: note: method 'runFrom:to:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'endDataTransfer' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:85:1: note: method 'endDataTransfer' declared here
- (void)endDataTransfer;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'clearExternalDragInformation' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:86:1: note: method 'clearExternalDragInformation' declared here
- (void)clearExternalDragInformation;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebView:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:94:1: note: method 'initWithWebView:' declared here
- (instancetype)initWithWebView:(TestWKWebView *)webView;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:additionalItemRequestLocations:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:95:1: note: method 'runFrom:to:additionalItemRequestLocations:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation additionalItemRequestLocations:(ProgressToCGPointValueMap)additionalItemRequestLocations;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'ensureInputSession' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:96:1: note: method 'ensureInputSession' declared here
- (void)ensureInputSession;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'setExternalItemProviders:defaultDropPreviews:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:97:1: note: method 'setExternalItemProviders:defaultDropPreviews:' declared here
- (void)setExternalItemProviders:(NSArray<NSItemProvider *> *)itemProviders defaultDropPreviews:(NSArray<UITargetedDragPreview *> *)previews;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'containsDraggedType:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:145:1: note: method 'containsDraggedType:' declared here
- (BOOL)containsDraggedType:(NSString *)type;
^
10 errors generated.
** BUILD FAILED **
```

</details>

whsieh commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -1211,7 +1211,7 @@ static BOOL overrideIsInHardwareKeyboardMode()
         auto simulator = adoptNS([[DragAndDropSimulator alloc] initWithWebView:webView.get()]);
         [simulator setExternalItemProviders:@[ itemProvider.get() ]];
-        [simulator runFrom:CGPointMake(50, 50) to:CGPointMake(150, 50)];
+        [simulator runFromElement:@"output" toElement:@"droparea"];
```
Hm...since these are element IDs and the method takes a query selector now, I would expect that these need to be `@"#output"` and `@"#droparea"` respectively? (I had suggested making the helper take a query selector instead of an ID, so that it could be used for scenarios where you wanted to start or end drags on elements that didn't have an `id` attribute, e.g. the only `img` element in the document or something along those lines).

aprotyas replied:
They were being used as selectors

```js
let element = document.querySelector('#%@');
```

_but_ of course this is not apparent at the call-site, which is confusing!

Instead I've renamed all parameters to indicate that we're expecting selector queries and updated the call-site to read `@"#output"` and `@"#droparea"` respectively.

https://github.com/WebKit/WebKit/pull/13844/commits/e92cb7bf8985c7c8b09a62d159bf24619ec278dd

aprotyas commented on Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
```diff
@@ -146,4 +146,11 @@ typedef NSDictionary<NSNumber *, NSValue *> *ProgressToCGPointValueMap;
+- (void)runFromElement:(NSString *)startSelector toElement:(NSString *)endSelector;
+
+@end
+
```
Extra newline here. I'll remove it before asking to `merge-queue` this PR.

aprotyas replied:
Removed in https://github.com/WebKit/WebKit/commit/d9f8d6940b38b36eafdacb26de28cd7d1d62a331

aprotyas committed:
[ iOS simulator ] 2X TestWebKitAPI.DragAndDropTests.ExternalSourceDataTransferItemGet (API-Tests) are constant timeouts
https://bugs.webkit.org/show_bug.cgi?id=256746
rdar://109222780

Reviewed by Wenson Hsieh.

The API tests TestWebKitAPI.DragAndDropTests.ExternalSourceDataTransferItemGet[Folder,PlainTextFile]AsEntry
consistently time out when running on an iOS simulator simulating iPhone
12. These timeouts have surfaced recently given we bumped the default
iPhone simulator to iPhone 12 in 256495@main. When running on iPhone 12,
the change in screen resolution meant that some of our hardcoded drag
gestures became out-of-place, causing unwarranted behavior and leading
to the timeouts.

This commit hardens the test cases against similar changes in the future
by introducing a way to start a drag over an element in the DOM and
ending it at another element in the DOM. Doing so required imbibing
`TestWKWebView` with a `getElementMidpoint` method, which takes the
selector for a DOM element and returns its midpoint, and
`DragAndDropSimulator` with a `runFromElement:toElement:` method, which
is similar in spirit to the `runFrom:to:` except the drag source and
destination are now specified as DOM element selectors, rather than
locations.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm: Added.
(-[DragAndDropSimulator runFromElement:toElement:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView getElementMidpoint:]):

Signed-off-by: Abrar Rahman Protyasha <a_protyasha@apple.com>
Canonical link: https://commits.webkit.org/264080@main
